worlds/celeste64/Rules.py

```diff
@@ -1,104 +1,346 @@
+                              [ItemName.ground_dash, ItemName.air_dash, ItemName.skid_jump],
+                              [ItemName.feather, ItemName.traffic_block, ItemName.air_dash],
+                              [ItemName.traffic_block, ItemName.ground_dash, ItemName.air_dash]],
+    LocationName.badeline_2: [[ItemName.double_dash_refill, ItemName.feather, ItemName.traffic_block, ItemName.breakables, ItemName.air_dash, ItemName.climb],
+                              [ItemName.traffic_block, ItemName.air_dash, ItemName.skid_jump],
+                              [ItemName.ground_dash, ItemName.air_dash, ItemName.skid_jump],
+                              [ItemName.feather, ItemName.traffic_block, ItemName.air_dash],
+                              [ItemName.traffic_block, ItemName.ground_dash, ItemName.air_dash]],
+    LocationName.badeline_3: [[ItemName.double_dash_refill, ItemName.feather, ItemName.traffic_block, ItemName.breakables, ItemName.air_dash, ItemName.climb],
+                              [ItemName.traffic_block, ItemName.air_dash, ItemName.skid_jump],
+                              [ItemName.ground_dash, ItemName.air_dash, ItemName.skid_jump],
+                              [ItemName.feather, ItemName.traffic_block, ItemName.air_dash],
+                              [ItemName.traffic_block, ItemName.ground_dash, ItemName.air_dash]],
+
+    LocationName.sign_2: [[ItemName.breakables, ItemName.ground_dash],
+                          [ItemName.breakables, ItemName.air_dash]],
+    LocationName.sign_5: [[ItemName.double_dash_refill, ItemName.feather, ItemName.traffic_block, ItemName.breakables, ItemName.air_dash, ItemName.climb],
+                          [ItemName.traffic_block, ItemName.air_dash, ItemName.skid_jump],
+                          [ItemName.ground_dash, ItemName.air_dash, ItemName.skid_jump],
+                          [ItemName.feather, ItemName.traffic_block, ItemName.air_dash],
+                          [ItemName.traffic_block, ItemName.ground_dash, ItemName.air_dash]],
+
+    LocationName.car_2: [[ItemName.breakables, ItemName.ground_dash],
+                         [ItemName.breakables, ItemName.air_dash]],
+}
+
+
+def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
+    active_logic_mapping: Dict[str, List[List[str]]]
+
+    if world.options.logic_difficulty == "standard":
+        if world.options.move_shuffle:
+            active_logic_mapping = location_standard_moves_logic
+        else:
+            active_logic_mapping = location_standard_logic
+    else:
+        if world.options.move_shuffle:
+            active_logic_mapping = location_hard_moves_logic
+        else:
+            active_logic_mapping = location_hard_logic
+
+    if loc not in active_logic_mapping:
+        return True
+
+    for possible_access in active_logic_mapping[loc]:
+        if state.has_all(possible_access, world.player):
+            return True
+
+    return False
```

Comment:
For performance, I would very much prefer if you moved as many conditions out of the actual access rule evaluation as possible, and did them ahead of time.
This if/else is going to be called each time an item is placed, so 1000s or 10000s of times, when it could be done once 

The simplest change I can think of is moving `active_location_mapping` to an attribute of e.g. `Celeste64World` that you set once.
So, move this whole if/else block somewhere else and save the attribute to the world (PyCharm might complain about "setting attribute outside of `__init__`, maybe you can figure something out)

```
def set_rules(world: Celeste64World):
    if world.options.logic_difficulty == "standard":
        if world.options.move_shuffle:
            world.active_logic_mapping = location_standard_moves_logic
        else:
            world.active_logic_mapping = location_standard_logic
    else:
        if world.options.move_shuffle:
            world.active_logic_mapping = location_hard_moves_logic
        else:
            world.active_logic_mapping = location_hard_logic

    ...
```

and then in location_rule, you can just have:

```
def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
    if loc not in world.active_logic_mapping:
        return True

    for possible_access in world.active_logic_mapping[loc]:
        if state.has_all(possible_access, world.player):
            return True

    return False
```

At this point, I'd be willing to approve the PR (after reviewing everything else as well)
But, there are more optimisations to be made that I would recommend that I will detail here as well.

Now that we have the active location mapping at an earlier point, instead of this "in" check inside the location_rule function, you could just do

```
def set_rules(world: Celeste64World):
    for location in world.multiworld.get_locations(world.player):
        if location in world.active_location_mapping:
            set_rule(location, lambda state, location=location: location_rule(state, world, location.name))
        else:
            set_rule(location, lambda state: True)
```

With location_rule now being:

```
def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
    for possible_access in world.active_logic_mapping[loc]:
        if state.has_all(possible_access, world.player):
            return True

    return False
```

And finally, we can replace this early exiting for loop with an `any` (which also inherently exits early)

```
def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
    return any(
        state.has_all(possible_access, world.player)
        for possible_access in world.active_logic_mapping[loc]
   )
```

Reply:
I'll give my thoughts on that, sure. Let me put the "size of the game" thing aside for a second, I'll come back to it.

So, I've implemented the changes I suggested and I'm observing about a 30% speedup. It's not crazy good or anything, but considering:
1. These suggestions are fairly simple (in the sense that they don't increase overall line count and still keep all of the same code, just shuffled around)
2. this is a *change* you're making from the original implementation (i.e. it is "in scope" of the PR)
3. "Avoid option checking inside of access rules" **is** a common piece of advice given by multiple world and core maintainers, like I'm not just pulling this out of my backside, there is some "consensus" around this - I disagree with the wording choice of "minutia" here because of this

It is something I will feel inclined to comment on.

Now, the size of the game might be a good argument, and I think I'm "emotionally" willing to let it go? But **please please please** consider the first suggestion I made, it is very simple and already gets 25% of a speedup, everything else is in the 1%-2% range.

I will however say that I'm not sure I'm **generically** comfortable with the idea - and you didn't say this, but this would be my takeaway from this conversation - that I should take the size of a game into consideration when reviewing PRs like this (new game or big systems refactor). This problem is something I'd absolutely be calling out and not letting fly for a 500 location game - Is it that fair or reasonable for me to be softer on it when the game is smaller? I'm not sure about that